### PR TITLE
Datadog Connector Component

### PR DESCRIPTION
**Pull request by gord02**
This pull request is for a newly created Datadog connector that replaces the need for the Datadog processor. This replacement is needed because the processor made use of a work-around that directly allowed the processor to export data. This is not something that adheres to the OpenTelemetry standard and thus the previous system should be deprecated and replaced by this new connector.

Link to issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19740

### CODE REVIEW

File: connector/datadogconnector/factory.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// defines the consumer type of the connector
+// we want to consume traces and export metrics therefore define nextConsumer as metrics, consumer is the next component in the pipeline
+func createTracesToMetricsConnector(_ context.Context, params connector.CreateSettings, cfg component.Config, nextConsumer consumer.Metrics) (connector.Traces, error) {
+	c, err := newConnector(params.Logger, cfg)
+	if err != nil {
+		return nil, err
+	}
+	c.metricsConsumer = nextConsumer
```
Comment (mx-psi) [inline connector/datadogconnector/factory.go:39]: Instead of doing this, why not pass it as an argument to `newConnector`?

File: connector/datadogconnector/go.mod [hunk 2]
```diff
@@ -0,0 +1,88 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector
+
+go 1.20
+
+require (
+	github.com/DataDog/datadog-agent/pkg/proto v0.48.0-beta.1
```
Comment (mx-psi) [inline connector/datadogconnector/go.mod:6]: @songy23 can you check what version should we use here and if we need replace statements?

File: connector/datadogconnector/go.mod [hunk 3]
```diff
@@ -0,0 +1,88 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector
+
+go 1.20
```
Comment (mx-psi) [inline connector/datadogconnector/go.mod:3]: We are still using Go 1.19, this will be changed when #25116 is addressed

```suggestion
go 1.19
```

File: go.mod [hunk 4]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/cassandraexporter v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/clickhouseexporter v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/coralogixexporter v0.82.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter v0.82.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter v0.0.0-00010101000000-000000000000
```
Comment (mx-psi) [inline go.mod:17]: You can revert this change, it should stay as it is

File: go.mod [hunk 5]
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/attributesprocessor v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor v0.82.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/processor/datadogprocessor v0.82.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/datadogprocessor v0.0.0-00010101000000-000000000000
```
Comment (mx-psi) [inline go.mod:69]: ditto, revert this change

File: connector/datadogconnector/go.mod (hunk 2 above)
Comment (songy23) [reply connector/datadogconnector/go.mod:6]: We should use `v0.48.0-beta.1` , if the new component is included in dependabot config then we need replace as well.

File: connector/datadogconnector/go.mod [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	golang.org/x/sys v0.10.0 // indirect
+	golang.org/x/text v0.11.0 // indirect
+	golang.org/x/time v0.3.0 // indirect
+	golang.org/x/tools v0.9.1 // indirect
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20230629202037-9506855d4529 // indirect
+	google.golang.org/grpc v1.57.0 // indirect
+	google.golang.org/protobuf v1.31.0 // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+)
+
```
Comment (songy23) [inline connector/datadogconnector/go.mod:87]: Please add these 2 lines:
```
replace github.com/DataDog/datadog-agent/pkg/proto => github.com/DataDog/datadog-agent/pkg/proto v0.48.0-beta.1

replace github.com/DataDog/datadog-agent/pkg/trace => github.com/DataDog/datadog-agent/pkg/trace v0.48.0-beta.1
```

**Review by songy23 — COMMENTED** [review]
Is this ready for review? 

Also you need a rebase

File: connector/datadogconnector/metadata.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+type: datadog
+
+status:
+  class: connector
+  stability:
+    alpha: [traces_to_metrics]
+  distributions: [contrib]
```
Comment (mx-psi) [inline connector/datadogconnector/metadata.yaml:7]: For now, this is not on the contrib distribution, let's leave this for after we make a PR on opentelemetry-collector-releases and actually add this to the contrib distro.
```suggestion
  distributions: []
```

File: connector/datadogconnector/connector.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog"
+)
+
+// connectorImp is the schema for connector
+type connectorImp struct {
+	metricsConsumer consumer.Metrics // the next component in the pipeline to ingest data after connector
+	logger          *zap.Logger
+	started         bool
```
Comment (mx-psi) [inline connector/datadogconnector/connector.go:23]: We don't actually read this value, only set it, so we can get rid of it
```suggestion
```

(and also remove other references below)

File: connector/datadogconnector/connector_test.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+	factory := NewFactory()
+
+	creationParams := connectortest.NewNopCreateSettings()
+	cfg := factory.CreateDefaultConfig().(*Config)
+
+	// Test
+	traceConnector, err := factory.CreateTracesToMetrics(context.Background(), creationParams, cfg, consumertest.NewNop())
+	smc := traceConnector.(*connectorImp)
+
+	assert.Nil(t, err)
```
Comment (mx-psi) [inline connector/datadogconnector/connector_test.go:30]: nit: the usual way of doing this is 
```suggestion
	assert.NoError(t, err)
```

this gives a nicer error message

File: connector/datadogconnector/connector_test.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+func TestNewConnector(t *testing.T) {
+
+	factory := NewFactory()
+
+	creationParams := connectortest.NewNopCreateSettings()
+	cfg := factory.CreateDefaultConfig().(*Config)
+
+	// Test
+	traceConnector, err := factory.CreateTracesToMetrics(context.Background(), creationParams, cfg, consumertest.NewNop())
+	smc := traceConnector.(*connectorImp)
```
Comment (mx-psi) [inline connector/datadogconnector/connector_test.go:28]: This will panic if `traceConnector` is not of that type, see https://go.dev/tour/methods/15. Instead of doing this, use the `ok` form, and check that it returned true:
```suggestion
	smc, ok := traceConnector.(*connectorImp)
	assert.True(t, ok)
```

File: connector/datadogconnector/connector_test.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector/connectortest"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+)
+
+var _ component.Component = (*connectorImp)(nil) // testing that the connectorImp properly implements the type Component interface
```
Comment (songy23) [inline connector/datadogconnector/connector_test.go:16]: This should be in `connector.go` instead

**Review by songy23 — COMMENTED** [review]
Please follow https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components, you're missing several steps like add to versions.yaml, add doc.go, make generate-gh-issue-templates, etc.

File: connector/datadogconnector/metadata.yaml (hunk 7 above)
Comment (songy23) [reply connector/datadogconnector/metadata.yaml:7]: This is not resolved.

File: versions.yaml [hunk 12]
```diff
@@ -16,6 +16,7 @@ module-sets:
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector
+      - github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector
```
Comment (songy23) [inline versions.yaml:19]: Please keep in alphabetic order.

File: cmd/otelcontribcol/builder-config.yaml [hunk 13]
```diff
@@ -201,6 +201,7 @@ connectors:
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector v0.82.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector v0.82.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector v0.82.0
+  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector v0.82.0
```
Comment (songy23) [inline cmd/otelcontribcol/builder-config.yaml:204]: Please keep in alphabetic order.

File: cmd/otelcontribcol/builder-config.yaml [hunk 14]
```diff
@@ -416,6 +417,7 @@ replaces:
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector => ../../connector/routingconnector
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector => ../../connector/servicegraphconnector
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector => ../../connector/spanmetricsconnector
+  - github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector => ../../connector/datadogconnector
```
Comment (songy23) [inline cmd/otelcontribcol/builder-config.yaml:420]: same

**Comment by songy23** [comment]
> Generated code is out of date, please run "make genotelcontribcol" and commit the changes in this PR.

**Review by songy23 — APPROVED** [review]

File: .chloggen/datadogconnector.yaml [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: This is a new component that computes Datadog APM Stats in the event that trace pipelines are sampled.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [19740]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (mx-psi) [inline .chloggen/datadogconnector.yaml:18]:
```suggestion
subtext: | 
  This component is meant to replace the Datadog processor
```
